### PR TITLE
Commit staging godeps

### PR DESCRIPTION
**Pull request by caesarxuchao**
**What this PR does / why we need it**:
Committed the client-go/1.4/Godeps/Godeps.json, so user can run `godep restore` after they `go get` the entire repo.

Added a Go program to modify the Godesp.json to make it usable.

FYI @thockin

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/31644)

### CODE REVIEW

**Comment by caesarxuchao** [comment]
@pwittrock I added the milestone because it's making [kubernetes/client-go](https://github.com/kubernetes/client-go) usable, and it doesn't affect the main repository so it's low risk.

File: staging/src/k8s.io/client-go/godeps-json-updater.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+	"fmt"
+	"os"
+	"strings"
+)
+
+var godepsFile string
+var clientRepoImportPath string
+
+func init() {
+	flag.StringVar(&godepsFile, "godeps-file", "", "absolute path to Godeps.json")
```
Comment (lavalamp) [inline staging/src/k8s.io/client-go/godeps-json-updater.go]: no init, just

```go
var (
  godepsFile = flag.String("godeps-file", "", "absolute path to Godeps.json")
  ...
)
```

File: staging/src/k8s.io/client-go/godeps-json-updater.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+	// e.g., k8s.io/client-go/1.4
```
Comment (lavalamp) [inline staging/src/k8s.io/client-go/godeps-json-updater.go]: include things like this in the help string :)

File: staging/src/k8s.io/client-go/godeps-json-updater.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+func main() {
+	flag.Parse()
+	var g Godeps
+	if len(godepsFile) == 0 {
+		panic("absolute ath to Godeps.json is required")
+	}
+	if len(clientRepoImportPath) == 0 {
+		panic("import path to a version of client-go is required")
+	}
+	//f, err := os.Open(path)
```
Comment (lavalamp) [inline staging/src/k8s.io/client-go/godeps-json-updater.go]: delete dead code

File: staging/src/k8s.io/client-go/godeps-json-updater.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+	f, err := os.OpenFile(godepsFile, os.O_RDWR, 0666)
+	if err != nil {
+		panic(err.Error())
```
Comment (lavalamp) [inline staging/src/k8s.io/client-go/godeps-json-updater.go]: I'd import the go log package so you can do `log.Fatalf("unable to open %q: %v", *godepsFile, err)`

File: staging/src/k8s.io/client-go/godeps-json-updater.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+	g.Deps = g.Deps[:i]
+	b, err := json.MarshalIndent(g, "", "\t")
+	if err != nil {
+		panic(err.Error())
+	}
+	n, err := f.WriteAt(append(b, '\n'), 0)
+	if err != nil {
+		panic(err.Error())
+	}
+	f.Truncate(int64(n))
```
Comment (lavalamp) [inline staging/src/k8s.io/client-go/godeps-json-updater.go]: is there an error to check?

**Comment by lavalamp** [comment]
minor comments.

File: staging/src/k8s.io/client-go/godeps-json-updater.go (hunk 4 above)
Comment (caesarxuchao) [inline staging/src/k8s.io/client-go/godeps-json-updater.go]: I tried not to create another vendor directory to avoid any confusion. What do you think?

**Comment by thockin** [comment]
The real question is whether this (a lib) should vendor at all.  The go community at large says no.  This would mean that it should work against HEAD of all the (transitive) deps.  10 bucks says that doesn't play out.  The alternative is that consumers of this lib may end up double-importing things.

It's sort of lose-lose right now...  @jfrazelle since Go vendoring committee :)

**Comment by caesarxuchao** [comment]
@thockin can we ask our user to run godep? If they vendor client-go, and uses godep, they won't have double-importing, because `godep save` will flatten the dependencies. However, I don't know what will happen if they depends on a different version of a client-go dependency.

**Comment by jfrazelle** [comment]
So we have a user goal to have multiple versions of deps work so no promises but the concern is there and pretty clear :)

**Comment by caesarxuchao** [comment]
Discussed offline. The long term goal is to get rid of the vendored dependency in client-go. For now, checking in the Godeps.json will at least help users to get the right version of dependencies, if users don't vendor conflicting versions, so this PR is a net gain.

File: staging/src/k8s.io/client-go/godeps-json-updater.go (hunk 4 above)
Comment (caesarxuchao) [inline staging/src/k8s.io/client-go/godeps-json-updater.go]: Sorry, I misread your comment as `glog`. Will do.

**Comment by caesarxuchao** [comment]
@lavalamp comments addressed. PTAL. Thanks!

**Comment by caesarxuchao** [comment]
Applied the label according comments.

**Comment by thockin** [comment]
I don't think that it is kosher to force users to use godep.

**Comment by thockin** [comment]
I don't really want to see an endorsement of multi-instance libs.  I did
the survey today, and I read between the lines, and that just feels like
giving up.  It won't work for any lib that gets exposed out of a package,
which bit me when someone used grpc...

**Comment by caesarxuchao** [comment]
This is how we suggest our user to solve the multi-instance problem: https://github.com/kubernetes/client-go/blob/master/README.md#dependency-management

They are not beautiful solutions. I think the end goal is getting rid of most third-party dependencies, I haven't looked into which dependencies are actually needed. @krousey has started some work to carve out Prometheus.

**Comment by thockin** [comment]
ACK that the answers are all bad for now.  The community best practice is
to not vendor in libs, but that is not realistic at this moment in time.

**Comment by jfrazelle** [comment]
So I looked through some of these (and obviously this can be handled
tangent to this PR) but for instance why is pflag part of a client library.
I think a lot of this could be cleaned up :)

**Comment by krousey** [comment]
@jfrazelle 2 reasons. The resource package that defines Quantity also defines a flag type for it. 

https://github.com/kubernetes/kubernetes/blob/master/pkg/api/resource/quantity.go defines a quantity flag. Not sure why it needs to be a pflag. Probably compatible if it just used the standard flag package.

https://github.com/kubernetes/kubernetes/blob/master/pkg/client/unversioned/clientcmd/overrides.go provide methods to bind struct variables to flags in flagsets. Again, could probably just use standard flag and be compatible with pflag in a higher level package.

If you want to fix either of these, I will gladly review.

**Comment by jfrazelle** [comment]
sweet!!!!

**Comment by thockin** [comment]
Yeah, these are a great GREAT opportunity to demonstrate when and how to
use interfaces and packages more carefully.  I would love to see us show
the solution at a k8s all-hands and explain the importance of it..

**Comment by lavalamp** [comment]
Please feel free to fix with liberal application of interfaces. This stuff
happens because of the monorepo.

**Pull request merged**